File: app/src/main/java/com/lostincontext/playlists/PlaylistsFragment.java
package com.lostincontext.playlists;

import android.app.Activity;
import android.content.Intent;
import android.content.res.Resources;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.lostincontext.PlaylistLauncher;
import com.lostincontext.R;
import com.lostincontext.commons.list.SpacesItemDecoration;
import com.lostincontext.data.playlist.Playlist;
import com.lostincontext.databinding.PlaylistsScreenFragmentBinding;

import java.util.List;


public class PlaylistsFragment extends Fragment implements PlaylistsContract.View {

    private PlaylistsContract.Presenter presenter;

    private PlaylistsAdapter adapter;


    public static PlaylistsFragment newInstance() {
        return new PlaylistsFragment();
    }


    public PlaylistsFragment() { }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        PlaylistsScreenFragmentBinding binding = DataBindingUtil.inflate(inflater,
                                                                         R.layout.playlists_screen_fragment,
                                                                         container,
                                                                         false);
        RecyclerView recyclerView = binding.recyclerView;
        Resources resources = getResources();
        final int span = resources.getInteger(R.integer.grid_span);
        GridLayoutManager layoutManager = new GridLayoutManager(getContext(), span);
        recyclerView.setLayoutManager(layoutManager);
        int space = resources.getDimensionPixelSize(R.dimen.grid_spacing);
        recyclerView.addItemDecoration(new SpacesItemDecoration(space, span));

        adapter = new PlaylistsAdapter(presenter, presenter);

        layoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override public int getSpanSize(int position) {
                switch (adapter.getCurrentState()) {
                    case LOADING:
                    case ERROR:
                    case EMPTY:
                        return span;

                    case CONTENT:
                        return 1;

                    default:
                        throw new RuntimeException("invalid state");
                }
            }
        });
        recyclerView.setAdapter(adapter);
        recyclerView.setHasFixedSize(true);
        return binding.getRoot();
    }


    @Override
    public void onResume() {
        super.onResume();
        presenter.start();
    }

    @Override public void setPresenter(PlaylistsContract.Presenter presenter) {
        this.presenter = presenter;
    }

    @Override public void setPlaylists(List<Playlist> playlists) {
        adapter.setPlaylists(playlists);
    }

    @Override public void openDeezerFor(Playlist playlist) {
        PlaylistLauncher launcher = new PlaylistLauncher();
        launcher.launchPlaylist(getContext(), playlist, false);
    }

    @Override public void returnResult(Playlist playlist) {
        Intent returnIntent = new Intent();
        returnIntent.putExtra("extra", "coin coin");
        returnIntent.putExtra(PlaylistsContract.EXTRA_PLAYLIST, playlist);
        FragmentActivity activity = getActivity();
        activity.setResult(Activity.RESULT_OK, returnIntent);
        activity.finish();
    }
}

File: app/src/main/java/com/lostincontext/playlists/PlaylistsFragment.kt
package com.lostincontext.playlists

import android.app.Activity
import android.content.Intent
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.GridLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.lostincontext.PlaylistLauncher
import com.lostincontext.R
import com.lostincontext.application.LostApplication
import com.lostincontext.commons.list.SpacesItemDecoration
import com.lostincontext.commons.list.StatefulAdapter.ContentState
import com.lostincontext.data.playlist.Playlist
import com.lostincontext.databinding.PlaylistsScreenFragmentBinding
import javax.inject.Inject


class PlaylistsFragment : Fragment(), PlaylistsContract.View {

    @Inject lateinit internal var presenter: PlaylistsPresenter

    private lateinit var adapter: PlaylistsAdapter


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val userId = activity.intent.getLongExtra(PlaylistsContract.EXTRA_USER_ID,
                                                  PlaylistsContract.NO_USER)

        DaggerPlaylistsComponent.builder()
                .playlistsPresenterModule(PlaylistsPresenterModule(this, userId))
                .applicationComponent((activity.application as LostApplication).appComponent)
                .build()
                .inject(this)
    }


    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val binding = DataBindingUtil.inflate<PlaylistsScreenFragmentBinding>(inflater,
                                                                              R.layout.playlists_screen_fragment,
                                                                              container,
                                                                              false)
        val recyclerView = binding.recyclerView
        val resources = resources
        val span = resources.getInteger(R.integer.grid_span)
        val layoutManager = GridLayoutManager(context, span)
        recyclerView.layoutManager = layoutManager
        val space = resources.getDimensionPixelSize(R.dimen.grid_spacing)
        recyclerView.addItemDecoration(SpacesItemDecoration(space, span))

        adapter = PlaylistsAdapter(presenter, presenter, presenter)
        layoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                when (adapter.currentState) {
                    ContentState.LOADING,
                    ContentState.ERROR,
                    ContentState.EMPTY -> return span

                    ContentState.CONTENT -> {
                        when (adapter.getContentItemViewType(position)) {
                            R.id.view_type_standard -> return 1
                            R.id.view_type_load_more -> return span
                            else -> throw RuntimeException("invalid state")
                        }
                    }

                    else -> throw RuntimeException("invalid state")
                }
            }
        }
        recyclerView.adapter = adapter
        recyclerView.setHasFixedSize(true)


        return binding.root
    }

    override fun onResume() {
        super.onResume()
        presenter.start()
    }

    override fun setPlaylists(playlists: List<Playlist>,
                              total: Int) {
        adapter.setPlaylists(playlists, total)
    }

    override fun displayMorePlaylists(playlists: List<Playlist>,
                                      total: Int,
                                      size: Int) {
        adapter.setPlaylists(playlists, total)
    }

    override fun swap(playlists: List<Playlist>,
                      hasNext: Boolean) {
        adapter.swap(playlists)
    }

    override fun setContentState(contentState: ContentState) {
        adapter.setCurrentState(contentState)
    }

    override fun openDeezerFor(playlist: Playlist) {
        val launcher = PlaylistLauncher()
        launcher.launchPlaylist(context, playlist, false)
    }

    override fun returnResult(playlist: Playlist) {
        val returnIntent = Intent()
        returnIntent.putExtra(PlaylistsContract.EXTRA_PLAYLIST, playlist)
        val activity = activity
        activity.setResult(Activity.RESULT_OK, returnIntent)
        activity.finish()
    }

    companion object {

        fun newInstance(): PlaylistsFragment {
            return PlaylistsFragment()
        }
    }
}
